### PR TITLE
[SPARK-40796][CONNECT][BUILD] Check the generated python protos in GitHub Actions

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
Check the generated python protos in GitHub Actions

### Why are the changes needed?
1, `generate_protos.sh`:
- add an optional argument to specify the output path;
- update `sed -i ''` commands to make it works on both MacOS and Linux;

2, `dev/check-codegen-python.py`:
- can run it locally to check whether the generated files are out of sync;
- add it to CI to detect inconsistent protos;

image

3, python lint
- after install `mypy-protobuf`, need to fix lint in `python/pyspark/sql/connect/plan.py`

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
manually tests and CI

### CODE REVIEW

File: connector/connect/dev/generate_protos.sh [hunk 1]
```diff
@@ -55,10 +70,13 @@ buf generate --debug -vvv
 for f in `find gen/proto/python -name "*.py*"`; do
   # First fix the imports.
   if [[ $f == *_pb2.py || $f == *_pb2_grpc.py ]]; then
-    sed -i '' -e 's/from spark.connect import/from pyspark.sql.connect.proto import/g' $f
```
Comment (zhengruifeng) [inline connector/connect/dev/generate_protos.sh:58]: `sed -i ''` doesn't work in Linux, so make this change

**Comment by zhengruifeng** [comment]
I manually modify `expressions_pb2.py` and then this check failed https://github.com/zhengruifeng/spark/actions/runs/3249624241/jobs/5332272955

File: dev/check-codegen-python.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        else:
+            fail(
+                "Generated codes for pyspark-connect are out of sync! "
+                "Please run ./connector/connect/dev/generate_protos.sh"
+            )
+
+
+check_connect_protos()
+
+# TODO: also check generated codes in pyspark-ml and pyspark-mllib
```
Comment (grundprinzip) [inline dev/check-codegen-python.py:81]:
```suggestion
# TODO: also check generated code in pyspark-ml and pyspark-mllib.
```

File: dev/check-codegen-python.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+def run_cmd(cmd):
+    print(f"RUN: {cmd}")
+    if isinstance(cmd, list):
+        return subprocess.check_output(cmd).decode("utf-8")
+    else:
+        return subprocess.check_output(cmd.split(" ")).decode("utf-8")
+
+
+def check_connect_protos():
+    print("Start checking the generated codes in pyspark-connect")
```
Comment (grundprinzip) [inline dev/check-codegen-python.py:47]:
```suggestion
    print("Start checking the generated codes in pyspark-connect.")
```

File: python/pyspark/sql/connect/plan.py [hunk 4]
```diff
@@ -322,11 +322,15 @@ def _convert_measure(
     ) -> proto.Aggregate.AggregateFunction:
         exp, fun = m
         measure = proto.Aggregate.AggregateFunction()
-        measure.function.name = fun
+        measure.function.name = fun  # type: ignore[attr-defined]
```
Comment (grundprinzip) [inline python/pyspark/sql/connect/plan.py:325]: Why do these show up as required now, even though we had mypy checks activated before>

**Review by grundprinzip — COMMENTED** [review]
Thanks for doing this. Only a minor number of nits.

File: python/pyspark/sql/connect/plan.py [hunk 5]
```diff
@@ -382,7 +386,9 @@ def plan(self, session: Optional["RemoteSparkSession"]) -> proto.Relation:
         rel = proto.Relation()
         rel.join.left.CopyFrom(self.left.plan(session))
         rel.join.right.CopyFrom(self.right.plan(session))
-        rel.join.on.CopyFrom(self.to_attr_or_expression(self.on, session))
+        rel.join.on.CopyFrom(  # type: ignore[attr-defined]
```
Comment (amaliujia) [inline python/pyspark/sql/connect/plan.py:389]: maybe different version of `mypy` so that you add to add this?

File: dev/check-codegen-python.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+        if len(result.funny_files) > 0:
+            print(f"Incomparable files: {result.funny_files}")
+            failed = True
+
+        if len(result.diff_files) > 0:
+            print(f"Different files: {result.diff_files}")
+            failed = True
+
+        if not failed:
```
Comment (amaliujia) [inline dev/check-codegen-python.py:70]: nit: 

I usually suggest to do 
```
if condition is true
then else 
```
so it is 
```
if failed:
else:
```

but this is minor.

**Comment by zhengruifeng** [comment]
installing `mypy-protobuf` doesn't change the versions of `mypy`,`mypy-extensions`,`pytest-mypy-plugins`, but it changes the behaviors:

```
(spark_test) ➜  spark git:(master) ✗ pip list | grep -i mypy
mypy                          0.920
mypy-extensions               0.4.3
pytest-mypy-plugins           1.9.3
(spark_test) ➜  spark git:(master) ✗ ./dev/lint-python
starting python compilation test...
python compilation succeeded.

starting black test...
black checks passed.

starting flake8 test...
flake8 checks passed.

starting mypy annotations test...
annotations failed mypy checks:
python/pyspark/pandas/frame.py:6371: error: unused "type: ignore" comment
python/pyspark/pandas/frame.py:6428: error: unused "type: ignore" comment
python/pyspark/pandas/frame.py:6430: error: unused "type: ignore" comment
python/pyspark/pandas/resample.py:104: error: unused "type: ignore" comment
python/pyspark/pandas/resample.py:137: error: unused "type: ignore" comment
Found 5 errors in 2 files (checked 366 source files)
1



(spark_test) ➜  spark git:(master) ✗ pip install mypy-protobuf
Collecting mypy-protobuf
  Using cached mypy_protobuf-3.3.0-py3-none-any.whl (16 kB)
Collecting types-protobuf>=3.19.12
  Using cached types_protobuf-3.20.4-py3-none-any.whl (60 kB)
Requirement already satisfied: protobuf>=3.19.4 in /Users/ruifeng.zheng/.dev/miniconda3/envs/spark_test/lib/python3.10/site-packages (from mypy-protobuf) (4.21.6)
Installing collected packages: types-protobuf, mypy-protobuf
Successfully installed mypy-protobuf-3.3.0 types-protobuf-3.20.4
(spark_test) ➜  spark git:(master) ✗ ./dev/lint-python
starting python compilation test...
python compilation succeeded.

starting black test...
black checks passed.

starting flake8 test...
flake8 checks passed.

starting mypy annotations test...
annotations failed mypy checks:
python/pyspark/pandas/frame.py:6371: error: unused "type: ignore" comment
python/pyspark/pandas/frame.py:6428: error: unused "type: ignore" comment
python/pyspark/pandas/frame.py:6430: error: unused "type: ignore" comment
python/pyspark/pandas/resample.py:104: error: unused "type: ignore" comment
python/pyspark/pandas/resample.py:137: error: unused "type: ignore" comment
python/pyspark/sql/connect/plan.py:325: error: "AggregateFunction" has no attribute "function"  [attr-defined]
python/pyspark/sql/connect/plan.py:327: error: "AggregateFunction" has no attribute "function"  [attr-defined]
python/pyspark/sql/connect/plan.py:329: error: "AggregateFunction" has no attribute "function"  [attr-defined]
python/pyspark/sql/connect/plan.py:338: error: "Aggregate" has no attribute "measures"  [attr-defined]
python/pyspark/sql/connect/plan.py:342: error: "Type[Aggregate]" has no attribute "GroupingSet"  [attr-defined]
python/pyspark/sql/connect/plan.py:344: error: "Aggregate" has no attribute "grouping_sets"  [attr-defined]
python/pyspark/sql/connect/plan.py:385: error: "Join" has no attribute "on"  [attr-defined]
Found 12 errors in 3 files (checked 366 source files)
1
(spark_test) ➜  spark git:(master) pip list | grep -i mypy
mypy                          0.920
mypy-extensions               0.4.3
mypy-protobuf                 3.3.0
pytest-mypy-plugins           1.9.3
```

File: dev/check-codegen-python.py (hunk 6 above)
Comment (zhengruifeng) [reply dev/check-codegen-python.py:70]: let me update it

**Comment by zhengruifeng** [comment]
protos in existing master are out of sync, and are successfully detected by this PR:
```
Start checking the generated codes in pyspark-connect.
RUN: /__w/spark/spark/connector/connect/dev/generate_protos.sh /tmp/tmpme0gkjlx
Different files: ['commands_pb2.py', 'commands_pb2.pyi']
Generated files for pyspark-connect are out of sync! Please run ./connector/connect/dev/generate_protos.sh
Error: Process completed with exit code 255.
```

So I think it's ready

**Comment by zhengruifeng** [comment]
cc @HyukjinKwon @amaliujia @grundprinzip  please take another look

**Review by amaliujia — COMMENTED** [review]
LGTM

Just some nits. It's ok merge this PR fast to solve the out of sync proto files first then optionally follow up on these items.

**Comment by zhengruifeng** [comment]
Let me merge it first to make sure connect protos are in sync

**Comment by zhengruifeng** [comment]
merged into master

File: .github/workflows/build_and_test.yml [hunk 7]
```diff
@@ -601,6 +607,8 @@ jobs:
       run: ./dev/lint-java
     - name: Python linter
       run: PYTHON_EXECUTABLE=python3.9 ./dev/lint-python
+    - name: Python code generation check
+      run: PATH=$PATH:$HOME/buf/bin PYTHON_EXECUTABLE=python3.9 ./dev/check-codegen-python.py
```
Comment (HyukjinKwon) [inline .github/workflows/build_and_test.yml:611]: Let's skip this test if the script doesn't exist. This workflow runs in branch-3.2 and 3.3 too (see https://github.com/apache/spark/actions/runs/3263093569/jobs/5361307965)

Comment (zhengruifeng) [reply .github/workflows/build_and_test.yml:611]: got it, let me fix it

**Pull request closed without merging**